common/fixation_entry.go

```diff
@@ -142,38 +241,149 @@ func (fs *FixationStore) AppendEntry(
+	case timerStaleEntry:
+		fs.deleteStaleEntries(ctx, safeIndex, block)
+	}
+}
+
+func (fs *FixationStore) updateFutureEntry(ctx sdk.Context, safeIndex string, block uint64) {
+	if block != uint64(ctx.BlockHeight()) {
+		panic(fmt.Sprintf("Future entry: future block %d != current block %d", block, ctx.BlockHeight()))
+	}
+
+	latestEntry, found := fs.getUnmarshaledEntryForBlock(ctx, safeIndex, block-1)
+	if found {
+		// if the latest entry has DeleteAt set, then transfer this DeleteAt to this
+		// future entry (becoming latest). also, drop the old timer, and start a new
+		// timer. (note: // deletion, if any, cannot be for the current block, since
+		// it would have been processed at the beginning of the block.
+
+		if latestEntry.HasDeleteAt() {
+			entry := fs.getEntry(ctx, safeIndex, block)
+			entry.DeleteAt = latestEntry.DeleteAt
+			latestEntry.DeleteAt = math.MaxUint64
+			fs.setEntry(ctx, entry)
+
+			if entry.IsDeletedBy(block) {
+				// if DeleteAt is exactly now as we transition from future entry to
+				// be the latest entry, then simply invoke deleteMarkedEntry().
+				key := encodeForTimer(latestEntry.Index, latestEntry.Block, timerDeleteEntry)
+				fs.tstore.DelTimerByBlockHeight(ctx, block, key)
+				fs.deleteMarkedEntry(ctx, safeIndex, block)
+			} else {
+				// otherwise, then replace the old timer (for the previous latest)
+				// with a new timer (for the new latest).
+				fs.replaceTimer(ctx, latestEntry, entry, timerDeleteEntry)
+			}
+		}
+
+		// latest entry had extra refcount for being the latest; so drop that refcount
+		// because from now on it is no longer so.
+
+		fs.putEntry(ctx, latestEntry)
+	}
+}
+
+func (fs *FixationStore) deleteMarkedEntry(ctx sdk.Context, safeIndex string, block uint64) {
+	entry := fs.getEntry(ctx, safeIndex, block)
+	blockHeight := uint64(ctx.BlockHeight())
+
+	if entry.DeleteAt != blockHeight {
+		panic(fmt.Sprintf("DelEntry entry deleted %d != current ctx block %d", entry.DeleteAt, uint64(ctx.BlockHeight())))
+	}
+
+	fs.setEntryIndex(ctx, safeIndex, false)
+	fs.putEntry(ctx, entry)
+
+	// forcefully remove all future entries: they were never referenced hence do not
+	// require stale-period.
+
+	store := fs.getEntryStore(ctx, safeIndex)
+
+	iterator := sdk.KVStoreReversePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	var entriesToRemove []types.Entry
+	for ; iterator.Valid(); iterator.Next() {
+		var entry types.Entry
+		fs.cdc.MustUnmarshal(iterator.Value(), &entry)
+
+		if entry.Block <= blockHeight {
+			break
+		}
+
+		entriesToRemove = append(entriesToRemove, entry)
+	}
+
+	for _, entry := range entriesToRemove {
+		key := encodeForTimer(entry.Index, entry.Block, timerFutureEntry)
+		fs.tstore.DelTimerByBlockHeight(ctx, entry.Block, key)
+		fs.removeEntry(ctx, entry.Index, entry.Block)
+	}
+}
+
+func (fs *FixationStore) deleteStaleEntries(ctx sdk.Context, safeIndex string, _ uint64) {
 	store := fs.getEntryStore(ctx, safeIndex)
+	block := uint64(ctx.BlockHeight())
```

Comment:
current blockheight member name is different for every methodm lets call it currentBlockHeight in all the methods